crates/bdk/src/wallet/mod.rs

```diff
@@ -2330,6 +2379,120 @@ impl<D> Wallet<D> {
+    where
+        D: PersistBackend<ChangeSet>,
+    {
+        let connected_to = match height.checked_sub(1) {
+            Some(prev_height) => BlockId {
+                height: prev_height,
+                hash: block.header.prev_blockhash,
+            },
+            None => BlockId {
+                height,
+                hash: block.block_hash(),
+            },
+        };
+        self.apply_block_connected_to(block, height, connected_to)
+            .map_err(|err| match err {
+                ApplyBlockError::CannotConnect(err) => err,
+                ApplyBlockError::UnexpectedConnectedToHash { .. } => {
+                    unreachable!("hash is derived from block directly")
+                }
+            })
+    }
+
+    /// Applies relevant transactions from `block` of `height` to the wallet, and connects the
+    /// block to the internal chain.
+    ///
+    /// The `connected_to` parameter informs the wallet how this block connects to the internal
+    /// [`LocalChain`]. Relevant transactions are filtered from the `block` and inserted into the
+    /// internal [`TxGraph`].
+    pub fn apply_block_connected_to(
+        &mut self,
+        block: &Block,
+        height: u32,
+        connected_to: BlockId,
+    ) -> Result<(), ApplyBlockError>
+    where
+        D: PersistBackend<ChangeSet>,
+    {
+        let block_hash = block.block_hash();
+        if connected_to.height == height && connected_to.hash != block_hash {
+            return Err(ApplyBlockError::UnexpectedConnectedToHash {
+                connected_to_hash: connected_to.hash,
+                expected_hash: block_hash,
+            });
+        }
+        if let Some(prev_height) = height.checked_sub(1) {
+            let prev_hash = block.header.prev_blockhash;
+            if prev_height == connected_to.height && prev_hash != connected_to.hash {
+                return Err(ApplyBlockError::UnexpectedConnectedToHash {
+                    connected_to_hash: connected_to.hash,
+                    expected_hash: prev_hash,
+                });
+            }
+        }
+
+        let chain_update = local_chain::Update {
+            tip: {
+                let this_block = BlockId {
+                    height,
+                    hash: block_hash,
+                };
+                let block_ids = match height {
+                    h if h > connected_to.height => vec![connected_to, this_block],
+                    h if h < connected_to.height => vec![this_block, connected_to],
+                    _ => vec![this_block],
+                };
+                let mut tip = Option::<CheckPoint>::None;
+                for id in block_ids {
+                    let new_tip = match tip {
+                        Some(prev_id) => prev_id.push(id).expect("must push"),
+                        None => CheckPoint::new(id),
+                    };
+                    tip = Some(new_tip);
+                }
+                tip.unwrap()
```

Comment:
```suggestion
let tip = block_ids[1..].iter().fold(CheckPoint::new(block_ids[0]), |tip, id| tip.push(id).expect("in order"))
```
Or something :^)

Can make a method on CheckPoint like `::from_iter` to reduce cognitive load.

Reply:
Addressed in fc2914af10aaca899af52d65e01d39c3d91ef8d5. I made a convenience method on `CheckPoint`.